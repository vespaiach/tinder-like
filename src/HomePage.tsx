import {
    Button,
    ButtonGroup,
    Card,
    CardActions,
    CardContent,
    CardMedia,
    CircularProgress,
    Typography,
} from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import { FavoriteRounded as FavoriteRoundedIcon } from '@material-ui/icons';
import Skeleton from '@material-ui/lab/Skeleton';

import BrokenHeartIcon from './BrokenHeartIcon';

const useStyles = makeStyles({
    img: {
        height: 320,
    },
    cardContent: {
        paddingLeft: 0,
        paddingRight: 0,
    },
    footerRoot: {
        paddingRight: 0,
        paddingLeft: 0,
    },
});

interface HomePageProps {
    id: string;
    name: string;
    picture: string;
    gender?: string | null;
    age?: number | null;
    loading: boolean;
    onLike: (id: string) => void;
    onDislike: (id: string) => void;
}
export default function HomePage({
    name,
    gender,
    age,
    picture,
    id,
    loading,
    onDislike,
    onLike,
}: HomePageProps) {
    const classes = useStyles();
    const loadingDetails = !gender || !age;
    return (
        <Card elevation={0}>
            {loading ? (
                <Skeleton animation="wave" variant="rect" height="320" className={classes.img} />
            ) : (
                <CardMedia
                    component="img"
                    alt={name}
                    className={classes.img}
                    image={picture}
                    title={name}
                />
            )}
            {loading ? (
                <CardContent className={classes.cardContent}>
                    <Skeleton height={68} />
                    <Skeleton />
                    <Skeleton />
                </CardContent>
            ) : (
                <CardContent className={classes.cardContent}>
                    <Typography gutterBottom variant="h5" component="h2" data-testid="username">
                        {name}
                    </Typography>
                    <Typography variant="body2" color="textSecondary" component="div" data-testid="userbio">
                        {loadingDetails ? (
                            <CircularProgress color="secondary" size={24} />
                        ) : (
                            <>
                                Sex: {gender}
                                <br />
                                Age: {age === null ? 'loadingDetails...' : age}
                            </>
                        )}
                    </Typography>
                </CardContent>
            )}
            {!loading && (
                <CardActions className={classes.footerRoot}>
                    <ButtonGroup disableElevation variant="contained" color="primary" fullWidth>
                        <Button
                            data-testid="btnDislike"
                            size="small"
                            color="primary"
                            startIcon={<BrokenHeartIcon title="Broken Heart" />}
                            onClick={() => onDislike(id)}>
                            Dislike
                        </Button>
                        <Button
                            data-testid="btnLike"
                            size="small"
                            color="primary"
                            startIcon={<FavoriteRoundedIcon />}
                            onClick={() => onLike(id)}>
                            Like
                        </Button>
                    </ButtonGroup>
                </CardActions>
            )}
        </Card>
    );
}
