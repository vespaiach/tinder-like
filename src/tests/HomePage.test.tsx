import { fireEvent, getByText, render } from '@testing-library/react';

import HomePage from '../HomePage';

describe('Test HomePage component', () => {
    let like: jest.Mock<any, any>;
    let dislike: jest.Mock<any, any>;

    beforeEach(() => {
        like = jest.fn();
        dislike = jest.fn();
    });

    test('should render skeleton loader', () => {
        const { container } = render(
            <HomePage
                loading
                id="1"
                name="Toan Nguyen"
                picture="picture"
                onLike={like}
                onDislike={dislike}
            />
        );
        expect(container).toMatchSnapshot();
    });

    test('should render profile correctly', () => {
        const { container } = render(
            <HomePage
                loading={false}
                id="1"
                name="Toan Nguyen"
                picture="picture"
                onLike={like}
                onDislike={dislike}
            />
        );
        expect(container).toMatchSnapshot();
    });

    test('should call like function', () => {
        const { container } = render(
            <HomePage
                loading={false}
                id="1"
                name="Toan Nguyen"
                picture="picture"
                onLike={like}
                onDislike={dislike}
            />
        );
        fireEvent.click(getByText(container, 'Like'));

        expect(like.mock.calls.length).toBe(1);
        expect(dislike.mock.calls.length).toBe(0);
        expect(like.mock.calls[0][0]).toBe('1');
    });

    test('should call dislike function', () => {
        const { container } = render(
            <HomePage
                loading={false}
                id="1"
                name="Toan Nguyen"
                picture="picture"
                onLike={like}
                onDislike={dislike}
            />
        );
        fireEvent.click(getByText(container, 'Dislike'));

        expect(dislike.mock.calls.length).toBe(1);
        expect(like.mock.calls.length).toBe(0);
        expect(dislike.mock.calls[0][0]).toBe('1');
    });
});
