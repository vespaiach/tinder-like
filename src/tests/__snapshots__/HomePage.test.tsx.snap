// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test HomePage component should render profile correctly 1`] = `
<div>
  <div
    class="MuiPaper-root MuiCard-root MuiPaper-elevation0 MuiPaper-rounded"
  >
    <img
      alt="Toan Nguyen"
      class="MuiCardMedia-root makeStyles-img-4 MuiCardMedia-media MuiCardMedia-img"
      src="picture"
      title="Toan Nguyen"
    />
    <div
      class="MuiCardContent-root makeStyles-cardContent-5"
    >
      <h2
        class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom"
        data-testid="username"
      >
        Toan Nguyen
      </h2>
      <div
        class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
        data-testid="userbio"
      >
        <div
          class="MuiCircularProgress-root MuiCircularProgress-colorSecondary MuiCircularProgress-indeterminate"
          role="progressbar"
          style="width: 24px; height: 24px;"
        >
          <svg
            class="MuiCircularProgress-svg"
            viewBox="22 22 44 44"
          >
            <circle
              class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
              cx="44"
              cy="44"
              fill="none"
              r="20.2"
              stroke-width="3.6"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="MuiCardActions-root makeStyles-footerRoot-6 MuiCardActions-spacing"
    >
      <div
        class="MuiButtonGroup-root MuiButtonGroup-fullWidth MuiButtonGroup-disableElevation MuiButtonGroup-contained"
        role="group"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall MuiButton-disableElevation MuiButton-fullWidth"
          data-testid="btnDislike"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <span
              class="MuiButton-startIcon MuiButton-iconSizeSmall"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <title>
                  Broken Heart
                </title>
                <path
                  d="M12,21.35L10.55,20.03C5.4,15.36 2,12.27 2,8.5C2,5.41 4.42,3 7.5,3C8.17,3 8.82,3.12 9.44,3.33L13,9.35L9,14.35L12,21.35V21.35M16.5,3C19.58,3 22,5.41 22,8.5C22,12.27 18.6,15.36 13.45,20.03L12,21.35L11,14.35L15.5,9.35L12.85,4.27C13.87,3.47 15.17,3 16.5,3Z"
                />
              </svg>
            </span>
            Dislike
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButtonGroup-grouped MuiButtonGroup-groupedHorizontal MuiButtonGroup-groupedContained MuiButtonGroup-groupedContainedHorizontal MuiButtonGroup-groupedContainedPrimary MuiButton-containedPrimary MuiButton-containedSizeSmall MuiButton-sizeSmall MuiButton-disableElevation MuiButton-fullWidth"
          data-testid="btnLike"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            <span
              class="MuiButton-startIcon MuiButton-iconSizeSmall"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M13.35 20.13c-.76.69-1.93.69-2.69-.01l-.11-.1C5.3 15.27 1.87 12.16 2 8.28c.06-1.7.93-3.33 2.34-4.29 2.64-1.8 5.9-.96 7.66 1.1 1.76-2.06 5.02-2.91 7.66-1.1 1.41.96 2.28 2.59 2.34 4.29.14 3.88-3.3 6.99-8.55 11.76l-.1.09z"
                />
              </svg>
            </span>
            Like
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Test HomePage component should render skeleton loader 1`] = `
<div>
  <div
    class="MuiPaper-root MuiCard-root MuiPaper-elevation0 MuiPaper-rounded"
  >
    <span
      class="MuiSkeleton-root MuiSkeleton-rect makeStyles-img-1 MuiSkeleton-wave"
    />
    <div
      class="MuiCardContent-root makeStyles-cardContent-2"
    >
      <span
        class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
        style="height: 68px;"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
      />
      <span
        class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
      />
    </div>
  </div>
</div>
`;
